MLTransform

AnandInguva opened the pull request:
Design doc: https://docs.google.com/document/d/1rQkSm_8tseLqDQaLohtlCGqt5pvMaP0XIpPi5UD0LCQ/edit

This PR introduces TFTProcessHandlerSchema which takes a pcoll and perform user provided data processing transforms. TFTProcessHandler uses tensorflow_transform's beam implementation for computing results. 

This PR aims to provide a simple to use API which abstracts the user away from the complexities that one can find with using tensorflow transform such as providing a tensorflow feature spec etc.


### Not considered/implemented in this PR
## Alternative design -  Alternative way to pass columns to the transforms

Instead of passing columns to the every data transform, we can let MLTransform take a param `columns` and assume that all the data transforms within the scope of that MLTransform are applicable only on the input `columns` passed.

But this could have a downside if the user calls multple `MLTransform`s, then there will be fusion break(due to tft internal implementation)

```
 with beam.Pipeline() as p:
      transforms = [ComputeAndApplyVocab(...), TFIDF(...)]
      result = (
          p
          | "Create" >> beam.Create(input_data)
          | beam.Map(lambda x: ComputeAndVocabBatchedInputType(**x)
                     ).with_output_types(ComputeAndVocabBatchedInputType)
          | "MLTransform" >>
          base.MLTransform(columns=['x'], transforms=transforms)
```


------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

damccorm commented on sdks/python/apache_beam/ml/transforms/handlers.py:
```diff
@@ -0,0 +1,406 @@
+                                          common_types.InstanceDictType]]]):
+  """
+    A PTransform that converts a collection of NamedTuples or Rows into a
+    collection of dictionaries.
```
This would benefit from a description of why we want to do this (to convert to an acceptable TFT input type).

damccorm commented on sdks/python/apache_beam/ml/transforms/handlers.py:
```diff
@@ -0,0 +1,406 @@
+    if is_container:
+      dtype = typing.get_args(typ)[0]
+      if len(typing.get_args(typ)) > 1 or typing.get_origin(dtype) == Union:
+        raise RuntimeError(
+            f"Incorrect type specifications in {typ} for column {col_name}. "
+            f"Please specify a single type.")
+      if dtype not in _default_type_to_tensor_type_map:
+        raise TypeError(
+            f"Unable to identify type: {dtype} specified on column: {col_name}"
+            f". Please specify a valid type.")
```
In our error cases, it would be good to be clear about what a valid type is

damccorm commented on sdks/python/apache_beam/ml/transforms/handlers.py:
```diff
@@ -0,0 +1,406 @@
+        raise RuntimeError(
+            f"Incorrect type specifications in {typ} for column {col_name}. "
+            f"Please specify a single type.")
+      if dtype not in _default_type_to_tensor_type_map:
+        raise TypeError(
```
Does this check need to be applied in the `else` case as well? If so, should it be pulled out of the if/else?

damccorm commented on sdks/python/apache_beam/ml/transforms/handlers.py:
```diff
@@ -0,0 +1,406 @@
+    fields = row_type._inner_types()
+    return Dict[str, Union[tuple(fields)]]
+
+  def _get_artifact_location(self, pipeline: beam.Pipeline):
```
Instead of dumping this in the staging directory or a temp directory, should we require users to provide an output directory? Presumably, users will want a well defined location for retrieving their artifacts

AnandInguva replied:
Added it as optional. If user doesn't provide it, I am falling back to this approach

robertwb replied:
I don't think the staging/temp directory makes sense as a default--if there are artifacts to be produced/consumed, this should be a required argument.

damccorm commented on sdks/python/apache_beam/ml/transforms/tft_transforms.py:
```diff
@@ -0,0 +1,301 @@
+    self.columns = columns
+    self._args = args
+    self._kwargs = kwargs
+    self.has_artifacts = False
```
Nit: we could probably get rid of this variable by just returning `None` from `get_analyzer_artifacts` and checking that. Would save some boilerplate (which could get lost) in this file.

damccorm commented on sdks/python/apache_beam/ml/transforms/tft_transforms.py:
```diff
@@ -0,0 +1,301 @@
+  def __call__(self, data):
+    return self.apply(data, *self._args, **self._kwargs)
+
+  def get_analyzer_artifacts(self, data, col_name):
```
There's a bunch of places types could be helpful in this file

AnandInguva replied:
Done

damccorm reviewed (commented):
Didn't do a thorough review, but the approach here looks pretty good to me

AnandInguva commented:
R: @tvalentyn

AnandInguva commented on sdks/python/apache_beam/ml/transforms/tft_transforms.py:
```diff
@@ -0,0 +1,340 @@
+#
```
I think a better name for this file would transforms/tft.py.

AnandInguva commented:
R: @robertwb

robertwb commented on sdks/python/apache_beam/ml/transforms/handlers.py:
```diff
@@ -0,0 +1,431 @@
+
+  def _fail_on_non_gloabl_window(self, pcoll):
+    window_fn = pcoll.windowing.windowfn
+    if not isinstance(window_fn, GlobalWindows):
```
Do we support non-trivial triggering? If not (e.g. we want to do a global combine) we should use `pcoll.windowing.is_default()`.

robertwb commented on sdks/python/apache_beam/ml/transforms/handlers.py:
```diff
@@ -0,0 +1,431 @@
+      self, pcoll: beam.PCollection[tft_process_handler_schema_input_type]
+  ) -> beam.PCollection[beam.Row]:
+
+    self._fail_on_non_gloabl_window(pcoll)
```
This should only be the case for artifact production. We want to support artifact consumption operations in streaming mode.

AnandInguva replied:
Some transforms provide artifacts such as `tft.scale_to_0_1`. Some transforms consume https://www.tensorflow.org/tfx/transform/api_docs/python/tft/scale_by_min_max which takes the artifacts produced. 

So should I just check the transforms the user provided and determine if we need to fail the pipeline or not?

robertwb replied:
I thought all transforms had two modes, a transform-producing and transform-consuming variant. E.g. tft.scale_by_min_max in "producing" mode does a pass over the data to figure out what the min and max of the column is (to properly scale the data), and in consuming mode takes the previously-computed min and max to scale the data.

iindyk replied:
tft has two sets of transforms: [analyzers](https://github.com/tensorflow/transform/blob/master/tensorflow_transform/analyzers.py) and [mappers](https://github.com/tensorflow/transform/blob/master/tensorflow_transform/mappers.py). Analyzers typically do reductions (i.e. produce "artifact"), mappers can be either pure (don't need a reduction) or requiring analyzers. Analyzers basically produce a "placeholder" that is placed into a produced tf graph and is refereed to in the mappers. So the first thing we do is run the analysis stage that substitutes these placeholders with the reduced value as a constant. The artifact that tft produces is basically the tf graph (with assets) with these placeholders substituted with constants produced during analysis stage.

The downstream streaming variant just applies these transformations as if the reduced values were constants in the produced graph.

iindyk replied:
tft.scale_by_min_max if used in the `preprocessing_fn` that is passed to `AnalyzeandTransformDataset` (or `AnalyzeDataset`) will produce the min and max values in the artifact. [Output artifact](https://github.com/tensorflow/transform/blob/e6e2a6890572287d7900a37c8c40e03826c78b41/tensorflow_transform/output_wrapper.py#L60) of such run when appliead to data will do the scaling with the previously produced values.

We have no out-of-the-box way to use the produced min and max in any different way (i.e. to do something other than scaling)

AnandInguva replied:
As a non tensorflow user, if I use `tft.scale_by_min_max` on dataset A and got `min` and `max` value of the dataset A after preprocessing the entire data.  Later if I want to use the same `tft.scale_by_min_max` on dataset B using the `min`, 'max' values I got from dataset A, there is no way we can plug the artifacts we produced  previously right?

AnandInguva replied:
@robertwb So we could ask the user about the artifact location. For preprocessing, we generate artifacts and store them in `artifact_location`. This is the `producing` stage.

For the consuming stage, we will ask the user to provide the same location, we will load the artifacts(transform_fn) and do the processing with produced artifacts instead of computing them again.

robertwb replied:
Correct. How do we distinguish between the two modes?

AnandInguva replied:
We ask for the `artifact_location`. Validate `artifact_location` using `tft.TFTransformOutput`. If it fails, we assume we should compute artifacts. If the validation is successful, we load the `artifacts` and plug them in to the pipeline.

We can also ask the users explicitly with input parameters to TFTProcessHandler.

robertwb replied:
This seems rather risky to infer. Consider the cases

(1) In iterating on my algorithm, I run over larger and larger portions of my actual data. All the constants are fixed on my first run and never learn about the new data.

(2) Due to corruption or some other mismatch, the desired artifact fails to load. I silently compute new artifacts and try to apply these different normalizations to a model trained with the prior normalizations.  

I think we need to be explicit here.

iindyk replied:
In the consuming mode users don't call `tft.scale_by_min_max` -- it is already embedded in the artifact with the scaling values. So the just load the whole artifact and apply it to the same (or a subset) of features and get the transformations in the consuming mode.

If you need actually need the min and max values (and not just the callable scaler):
Although TFT doesn't always store the reduction values separately (it stores quantiles and vocab separately), there's a pretty clear path of how that could be done, so if you think this is necessary - we would support a contribution.

If you want to avoid changes in TFT, there's a hacky workaround as well: e.g. for `tft.scale_by_min_max` you could apply the model to [-1, 1] getting [res0, res1] and solve the system of equations:
(-1 - min)/(max - min) = res0
(1 - min)/(max-min) = res1

robertwb commented on sdks/python/apache_beam/ml/transforms/tft_transforms.py:
```diff
@@ -0,0 +1,416 @@
+    return "compute_and_apply_vocabulary"
+
+
+class Scale_To_ZScore(TFTOperation):
```
Class names should generally be CamelCase, not Camel_And_SnakeCase.

robertwb commented on sdks/python/apache_beam/ml/transforms/tft_transforms.py:
```diff
@@ -0,0 +1,416 @@
+
+    Args:
+      columns: A list of column names to apply the transformation on.
+      elementwise: If True, the transformation is applied elementwise.
```
What does "elementwise" mean?

AnandInguva replied:
From TFT documentation on tft.scale_to_0_1

If true, scale each element of the tensor independently.

elementwise: IIUC, if true, it calculates scale_to_0_1 on entire dataset. if false, it calculates on individual tensors.

I am just passing it if any user might be interested.

robertwb commented on sdks/python/apache_beam/ml/transforms/tft_transforms_test.py:
```diff
@@ -0,0 +1,393 @@
+          | beam.Map(lambda x: ComputeAndVocabBatchedInputType(**x)
+                     ).with_output_types(ComputeAndVocabBatchedInputType)
+          | "MLTransform" >>
+          base.MLTransform(process_handler=process_handler).with_transform(
```
What does "process_handler" mean? Is it an abstraction we need to expose to the user?

AnandInguva replied:
ProcessHandler handles the process logic related to a respective framework. Here TFTProcessHandler implements TFT code(which uses tft_beam.AnalyzeAndTransformDataset). Different frameworks will require different implementations. 

As of now, I think it would be complicated to the users to implement a ProcessHandler since it requires a full pass on data to determine some artifacts(such a min, max) and this requires combiners. I think it would be better if we provide implementations of ProcessHandlers(TFT, scikit-learn etc) to the users and make ProcessHandler internal. What do you think?

robertwb replied:
Yes, I think we should make this internal rather than exposed to the user.

robertwb replied:
I meant why are we exposing the notion of ProcessHandlers to the user. But maybe this is a larger discussion.

AnandInguva replied:
we are creating some data transformations with TFT. To process data using these transformations, we need to implement few things such as preparing data, schema for TFT AnalyzeAndTransformDataset etc. All the logic related to TFT will be abstracted in the ProcessHandler (TFTProcessHandler).

Why ProcessHandler? TFT is a framework and it has its own semantics.  In the future if we want to create data processing transforms using JAX, we can create a JAXProcessHandler that handles jax related processing steps.

To differentiate between frameworks, users need to pass `process_handler`. If we are just implementing data transformations with TFT, then we could remove the param `process_handler` and use it internally

robertwb commented on sdks/python/apache_beam/ml/transforms/base.py:
```diff
@@ -0,0 +1,119 @@
+
+class MLTransform(beam.PTransform[beam.PCollection[ExampleT],
+                                  beam.PCollection[MLTransformOutputT]],
+                  Generic[ExampleT, MLTransformOutputT, ]):
```
Remove this trailing comma?

robertwb commented:
As discussed on the doc, let's avoid process_handler as a public part of the API or required argument.

AnandInguva commented:
@robertwb @damccorm I addressed most of the comments. PTAL. 

1. Added artifact consumption vs production mode.
2. Made TFTProcessHandler internal to the MLTransform

AnandInguva commented:
retest this please

damccorm commented on sdks/python/apache_beam/ml/transforms/base.py:
```diff
@@ -0,0 +1,165 @@
+    # avoid circular import
+    # pylint: disable=wrong-import-order, wrong-import-position
+    from apache_beam.ml.transforms.handlers import TFTProcessHandler
+    process_handler = TFTProcessHandler(
```
Might be worth dropping a TODO here. Eventually, we'll want to map transforms to process_handlers (so if you say `ComputeAndApplyVocabulary` we use the `TFTProcessHandler`, if you say `MyCoolJaxFn` we use `JaxProcessHandler`). Don't need to implement anything, but it would be good to be clear on our path forward there.

AnandInguva replied:
Done.

damccorm commented on sdks/python/apache_beam/ml/transforms/handlers.py:
```diff
@@ -0,0 +1,436 @@
+  def _fail_on_non_default_windowing(self, pcoll: beam.PCollection):
+    if not pcoll.windowing.is_default():
+      raise RuntimeError(
+          "TFTProcessHandler only supports GlobalWindows when producing "
```
Nit: errors shouldn't explicitly reference `TFTProcessHandler` anymore, instead we should provide the transform name where we ran into this (or just the list of transform names)

AnandInguva replied:
I changed this from TFTProcessHandler to MLTransform.

damccorm commented on sdks/python/apache_beam/ml/transforms/handlers.py:
```diff
@@ -0,0 +1,436 @@
+      if not os.path.exists(os.path.join(
+          self.artifact_location, RAW_DATA_METADATA_DIR, SCHEMA_FILE)):
+        raise FileNotFoundError(
+            "Raw data metadata not found at %s" %
```
We should be more descriptive in these errors (something along the lines of "you're running in consume mode" what that means, and "have you run this in produce mode?")

AnandInguva replied:
Changed wording

AnandInguva commented on sdks/python/tox.ini:
```diff
@@ -326,6 +326,12 @@ commands =
   # Run all DataFrame API unit tests
   bash {toxinidir}/scripts/run_pytest.sh {envname} 'apache_beam/dataframe'
 
+[testenv:py{38,39}-tft-113]
```
tfx_bsl only available in 3.8 and 3.9 python - https://pypi.org/project/tfx-bsl/#files

damccorm replied:
Right, I forgot this was known.

AnandInguva commented on sdks/python/apache_beam/ml/transforms/base.py:
```diff
@@ -0,0 +1,165 @@
+      artifact_location: str,
+      artifact_mode: str = ArtifactMode.PRODUCE,
+      transforms: Optional[List[BaseOperation]] = None,
+      is_input_record_batches: bool = False,
+      output_record_batches: bool = False,
```
info about record_batch lives within the context of TFTProcessHandler.  I don't like the idea of passing this arg via operations since in the operations, we don't use this arg anywhere in the operation. It would be just a different way of inferring this arg in TFTProcessHandler.

Alternative would be MLTransform would take a `config: Dict[str, Any]`, in which we will instruct users to pass certain kind of args such as `{'out_record_batch' : True}`, we could pass this `config` arg to the respective process handler and discard  irrelevant args to that ProcessHandler.

damccorm replied:
> we don't use this arg anywhere in the operation

I think from a user's point of view we use it in the first one; `is_input_record_batches=True` just means "I'm expecting that its ok to pass record batches into the first transform, right?

AnandInguva replied:
Right now, the columns are specified at operation level instead of transform level. The entry point for column x could be at the beginning but the entry point at column y could be in the middle of the list. User might pass `input_record_batch=True` to the y column if they doesn't understand exactly what we were instructing.

If we ask the  user to provide like this, I feel like it could get a little complicated

```
transforms = [
   op1(columns=['x']),
   op2(columns=['y', input_record_batch=True],
   op3(columns=['z'] 
]
```

We can also iterate on this in the v2 since I guess this needs another discussion and remove the option for Record batches for now. We would support Dict[str, Any] in V1.  what do you think?

damccorm replied:
I'm comfortable iterating on it in v2

AnandInguva commented on sdks/python/apache_beam/ml/transforms/handlers.py:
```diff
@@ -0,0 +1,436 @@
+    #  we will convert scalar values to list values and TFT will ouput
+    # numpy array all the time.
+    if not self.is_input_record_batches:
+      raw_data |= beam.ParDo(ConvertScalarValuesToListValues())
```
We need to know before pipeline runtime whether it is a record batch or dict to determine the right schema.

So during pipeline construction, I use this flag to construct respective schema. I don't think this can be inferable later

AnandInguva commented on sdks/python/apache_beam/ml/transforms/base.py:
```diff
@@ -0,0 +1,119 @@
+
+import apache_beam as beam
+
+# TODO: Abstract methods are not getting pickled with dill.
```
Not relevant anymore. I tried today and I wasn't able to reproduce it now

AnandInguva commented on sdks/python/apache_beam/ml/transforms/tft_transforms.py:
```diff
@@ -0,0 +1,442 @@
+See the documentation for beam.ml.MLTransform for more details.
+
+Since the transforms in this module are implemented using TFT, they
+should be wrapped inside a TFTProcessHandler object before being passed
+to the beam.ml.MLTransform class. The TFTProcessHandler will let MLTransform
+know which type of input is expected and infers the relevant schema required
+for the TFT library.
```
Updated

AnandInguva commented on sdks/python/apache_beam/ml/transforms/base.py:
```diff
@@ -0,0 +1,119 @@
+ExampleT = TypeVar('ExampleT')
+MLTransformOutputT = TypeVar('MLTransformOutputT')
+
+# Input to the process data. This could be same or different from ExampleT.
```
Changed it

AnandInguva commented:
@damccorm  removed pyarrow record batch. Addressed other comments as well. PTAL


ref: https://github.com/apache/beam/issues/26640#issuecomment-1614764518

damccorm reviewed (approved):
I think this LGTM for our initial pass and we can follow up on everything else in future PRs. Known work that remains that I'm aware of:

1. Figure out how to support record batch types (and more generally, ProcessHandler specific parameters)
2. Follow up with tfx_bsl team to get 3.10+ support
3. (minor) rename tft_transforms.py to tft.py
4. Support remaining operations
5. Future work from https://docs.google.com/document/d/1rQkSm_8tseLqDQaLohtlCGqt5pvMaP0XIpPi5UD0LCQ/edit#heading=h.2jxsxqh

Lets sure we are tracking all of those somewhere (could be an issue or a doc, please just share whatever you choose)

AnandInguva commented:
R: @rszper

AnandInguva commented:
> R: @rszper

Actually, I am thinking I will iterate on the docs in the following PRs. Docs will include pydocs edits and doc pages as well for users
